File: tests/host/test_ssh_deadline.c
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include "sshmgr/ssh.h"
#include "util/proc.h"

static int completed(int rc, void *unused)
{
    (void)unused;
    assert(rc == PROC_RUN_TIMEOUT || rc == 255);
    puts("completion observed failure");
    fflush(stdout);
    return 0;
}

int main(int argc, char **argv)
{
    assert(argc == 3);
    struct profile p = {0};
    snprintf(p.dir, sizeof(p.dir), "%s", argv[1]);
    const char *op = argv[2];
    if (strcmp(op, "alive") == 0)
        assert(ssh_master_alive(&p) == -1);
    else if (strcmp(op, "exit") == 0)
        ssh_master_exit(&p);
    else if (strcmp(op, "start") == 0)
        assert(ssh_master_start(&p, "127.0.0.1", 1) == -1);
    else if (strcmp(op, "forward") == 0) {
        int rc = ssh_forward_add_tcp_observed(&p, "127.0.0.1", "127.0.0.1",
            54321, "127.0.0.1", 54321, completed, NULL);
        assert(rc == PROC_RUN_TIMEOUT || rc == 255);
    }
    else if (strcmp(op, "cancel") == 0) {
        int rc = ssh_forward_cancel_tcp(&p, "127.0.0.1", "127.0.0.1",
            54321, "127.0.0.1", 54321);
        assert(rc == PROC_RUN_TIMEOUT || rc == 255);
    }
    else {
        assert(strcmp(op, "exec") == 0);
        const char *command[] = {"sudo", "systemctl", "poweroff", NULL};
        assert(ssh_exec_bounded(&p, "127.0.0.1", command, 200) == PROC_RUN_TIMEOUT);
    }
    return 0;
}

File: tests/host/test_ssh_deadline.py
#!/usr/bin/env python3
"""A real OpenSSH client must not wait forever on an unresponsive master."""
import os
from pathlib import Path
import socket
import subprocess
import tempfile
import threading
import time

binary = Path(os.environ.get('SSH_DEADLINE_TEST', 'build/tests/test_ssh_deadline')).resolve()
for operation, budget in [('alive', 5), ('exit', 5), ('start', 1),
                           ('forward', 5), ('cancel', 5), ('exec', .2)]:
    with tempfile.TemporaryDirectory(prefix='hamn-ssh-', dir='/tmp') as directory:
        server = socket.socket(socket.AF_UNIX)
        server.bind(directory + '/ssh.sock')
        server.listen(1)
        server.settimeout(10)
        accepted, release = threading.Event(), threading.Event()
        errors = []

        def serve():
            try:
                with server.accept()[0]:
                    accepted.set()
                    assert release.wait(12)
            except Exception as error:
                errors.append(error)

        thread = threading.Thread(target=serve)
        thread.start()
        try:
            started = time.monotonic()
            result = subprocess.run([binary, directory, operation],
                capture_output=True, env=dict(os.environ, PATH='/usr/bin:/bin'), timeout=budget + 4)
            elapsed = time.monotonic() - started
            assert accepted.is_set(), (operation, result)
            assert result.returncode == 0, (operation, result)
            assert budget <= elapsed < budget + 4, (operation, elapsed)
            if operation == 'forward':
                assert result.stdout.count(b'completion observed failure') == 1, result
        finally:
            release.set()
            thread.join(timeout=12)
            server.close()
        assert not errors and not thread.is_alive(), errors
print('PASS: real SSH check/exit/start/forward/cancel/exec obey operation deadlines')
